src/resultAsync.ts

```diff
@@ -19,10 +19,86 @@ import {
   Some,
 } from './utilities.js';
 
+/**
+ * Allows to extract the Value of the given "operation"
+ * An operation can be a Result, ResultAsync or a Promise
+ *
+ * @example ResultValueOf<Result<string>> => string
+ */
+export type OperationValue<T> = T extends ResultAsync<infer TResultAsyncValue>
+  ? TResultAsyncValue
+  : T extends Promise<infer TPromiseValue>
+  ? TPromiseValue
+  : T extends Result<infer TResultValue>
+  ? TResultValue
+  : never;
+
 /**
  * Represents and asynchronous Result that could succeed with a value or fail with an error
  */
 export class ResultAsync<TValue = Unit, TError = string> {
+  /**
+   * Combines several results (and any error messages) into a single result.
+   * The returned result will be a failure if any of the input results are failures.
+   *
+   * @param record The Results to be combined.
+   * @returns A Result that is a success when all the input results are also successes.
+   */
+  static combine<
+    TOperationRecord extends Record<
+      string,
+      Result<unknown> | ResultAsync<unknown> | Promise<unknown>
+    >
+  >(
+    record: TOperationRecord
+  ): ResultAsync<{
+    [K in keyof TOperationRecord]: OperationValue<TOperationRecord[K]>;
+  }> {
+    const promises = Object.entries(record).map(([key, resultOrPromise]) =>
+      resultOrPromise instanceof Result
+        ? Promise.resolve(resultOrPromise)
+        : resultOrPromise instanceof ResultAsync
+        ? resultOrPromise.toPromise()
+        : resultOrPromise
+            .then((v) => Result.success(v))
+            .catch((e) =>
+              Result.failure(e instanceof Error ? e.message : 'Unknown error')
```

Comment:
@seangwright Currently, combineAsync allows passing Promises directly. I started with this idea, because I thought it lowers the bar to use this feature.

The trade-off is, that we need to guess the failure message, which also might be a technical error.
So I am mixing the semantics of error and failure.

What do you think about it?
I guess, we should *not allow* passing a Promise.
A Promise could be turned into a `ResultAsync` specifying the correct failure message.